File: Modules/BuiltInAlgorithms/Sources/SplaySort.swift
import AlgorithmKit
import SortEngineKit

/// An adaptive sort built on a splay tree — a self-balancing binary search tree that moves
/// whatever node was just accessed up to the root via a sequence of rotations (`leftRotate`/
/// `rightRotate`). Each element is inserted by first splaying the existing tree toward that
/// element's own key (`splay`, a combined search-and-rotate descent rather than a plain search),
/// which brings whatever key is closest to the new one to the root; the new node then simply
/// swaps in as the root, taking over one side of the just-splayed node's children and leaving the
/// other side attached to it. Every key here is a genuine held copy (`Node.key`), not an index back
/// into the array — unlike `TreeSort`, this insertion order matters for the tree's shape, so the
/// values themselves have to travel with each node.
///
/// Splay trees are self-adjusting: repeatedly accessing nearby keys keeps the tree shallow for
/// exactly that stretch, which is what gives this sort its adaptive character — an input that's
/// already sorted or nearly so inserts each new (extreme) key in roughly constant amortized time,
/// while a genuinely random insertion order costs the same amortized `O(log n)` per insertion an
/// ordinary balanced tree would. A final in-order traversal writes each key directly back into the
/// real array as it's visited, without a separate temporary buffer.
public struct SplaySort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "splaysort")
  public let metadata = AlgorithmMetadata(
    displayName: "Splay Sort",
    category: .insertion,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 304, coefficients: [293.481, 0.89195],
      measuredSafeCeiling: nil),
    stable: true,
    timeComplexity: ComplexityBounds(best: "O(n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(n)",
    iconName: "rotate.3d"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }

    final class Node {
      var key: Int
      var left: Node?
      var right: Node?
      init(_ key: Int) { self.key = key }
    }

    func leftRotate(_ x: Node) -> Node {
      let y = x.right!
      x.right = y.left
      y.left = x
      return y
    }

    func rightRotate(_ x: Node) -> Node {
      let y = x.left!
      x.left = y.right
      y.right = x
      return y
    }

    func splay(_ rootArg: Node?, _ key: Int) -> Node? {
      guard var root = rootArg else { return nil }

      if root.key > key {
        guard let left = root.left else { return root }
        if left.key > key {
          left.left = splay(left.left, key)
          root = rightRotate(root)
        } else {
          left.right = splay(left.right, key)
          if left.right != nil {
            root.left = leftRotate(left)
          }
        }
        return root.left == nil ? root : rightRotate(root)
      } else {
        guard let right = root.right else { return root }
        if right.key > key {
          right.left = splay(right.left, key)
          if right.left != nil {
            root.right = rightRotate(right)
          }
        } else {
          right.right = splay(right.right, key)
          root = leftRotate(root)
        }
        return root.right == nil ? root : leftRotate(root)
      }
    }

    func insertRec(_ rootArg: Node?, _ key: Int) -> Node {
      guard let splayed = splay(rootArg, key) else {
        return Node(key)
      }
      let inserted = Node(key)
      if splayed.key > key {
        inserted.right = splayed
        inserted.left = splayed.left
        splayed.left = nil
      } else {
        inserted.left = splayed
        inserted.right = splayed.right
        splayed.right = nil
      }
      return inserted
    }

    var root: Node?
    for i in 0..<n {
      root = insertRec(root, engine.values[i])
    }

    var index = 0
    func traverse(_ node: Node?) {
      guard let node else { return }
      traverse(node.left)
      engine.setValue(index, node.key)
      index += 1
      traverse(node.right)
    }
    traverse(root)
  }
}

File: App/Resources/AlgorithmDetails/splaysort/splaysort.java
import java.util.Arrays;

public class splaysort {
  private static int idx;

  private static class Node {
    int key;
    Node left;
    Node right;

    Node(int key) {
      this.key = key;
    }
  }

  private static Node leftRotate(Node x) {
    Node y = x.right;
    x.right = y.left;
    y.left = x;
    return y;
  }

  private static Node rightRotate(Node x) {
    Node y = x.left;
    x.left = y.right;
    y.right = x;
    return y;
  }

  private static Node splay(Node root, int key) {
    if (root == null) {
      return root;
    }
    if (root.key > key) {
      if (root.left == null) {
        return root;
      }
      if (root.left.key > key) {
        root.left.left = splay(root.left.left, key);
        root = rightRotate(root);
      } else {
        root.left.right = splay(root.left.right, key);
        if (root.left.right != null) {
          root.left = leftRotate(root.left);
        }
      }
      return root.left == null ? root : rightRotate(root);
    } else {
      if (root.right == null) {
        return root;
      }
      if (root.right.key > key) {
        root.right.left = splay(root.right.left, key);
        if (root.right.left != null) {
          root.right = rightRotate(root.right);
        }
      } else {
        root.right.right = splay(root.right.right, key);
        root = leftRotate(root);
      }
      return root.right == null ? root : leftRotate(root);
    }
  }

  private static Node insertRec(Node root, int key) {
    if (root == null) {
      return new Node(key);
    }
    root = splay(root, key);
    Node n = new Node(key);
    if (root.key > key) {
      n.right = root;
      n.left = root.left;
      root.left = null;
    } else {
      n.left = root;
      n.right = root.right;
      root.right = null;
    }
    return n;
  }

  private static void traverse(Node node, int[] result) {
    if (node != null) {
      traverse(node.left, result);
      result[idx++] = node.key;
      traverse(node.right, result);
    }
  }

  public static void sort(int[] arr) {
    int n = arr.length;
    Node root = null;
    for (int x : arr) {
      root = insertRec(root, x);
    }
    int[] result = new int[n];
    idx = 0;
    traverse(root, result);
    System.arraycopy(result, 0, arr, 0, n);
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
